atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/impl/KvTableMappingService.java

```diff
@@ -110,35 +130,136 @@ private TableReference getAlreadyExistingMappedTableName(TableReference tableRef
-        Cell key = Cell.create(getBytesForTableRef(tableRef), AtlasDbConstants.NAMESPACE_SHORT_COLUMN_BYTES);
-        if (kv.getAllTableNames().contains(AtlasDbConstants.NAMESPACE_TABLE)) {
-            kv.delete(AtlasDbConstants.NAMESPACE_TABLE, ImmutableMultimap.of(key, 0L));
+        removeTables(ImmutableSet.of(tableRef));
+    }
+
+    @Override
+    public void removeTables(Set<TableReference> tableRefs) {
+        if (kvs.getAllTableNames().contains(AtlasDbConstants.NAMESPACE_TABLE)) {
+            Multimap<Cell, Long> deletionMap = tableRefs.stream()
+                    .map(this::getKeyCellForTable)
+                    .collect(Multimaps.toMultimap(
+                            key -> key,
+                            unused -> AtlasDbConstants.TRANSACTION_TS,
+                            MultimapBuilder.hashKeys().arrayListValues()::build));
+            kvs.delete(AtlasDbConstants.NAMESPACE_TABLE, deletionMap);
+        }
+
+        // Need to invalidate the table refs in case we end up re-creating the same table again.
+        updateTableMapWith(() -> tableMapWithInvalidatedTableRefs(tableRefs));
+    }
+
+    private BiMap<TableReference, TableReference> tableMapWithInvalidatedTableRefs(Set<TableReference> tableRefs) {
+        BiMap<TableReference, TableReference> newMap = HashBiMap.create(tableMap.get());
+        tableRefs.forEach(newMap::remove);
+        return newMap;
+    }
+
+    @Override
+    public TableReference getMappedTableName(TableReference tableRef) throws TableMappingNotFoundException {
+        if (tableRef.getNamespace().isEmptyNamespace()) {
+            return tableRef;
+        }
+
+        TableReference shortTableName = getMappedTableRef(tableRef);
+        validateShortName(tableRef, shortTableName);
+        return shortTableName;
+    }
+
+    private TableReference getMappedTableRef(TableReference tableRef)
+            throws TableMappingNotFoundException {
+        TableReference candidate = tableMap.get().get(tableRef);
+        if (candidate == null) {
+            updateTableMap();
+            candidate = tableMap.get().get(tableRef);
+            if (candidate == null) {
+                throw new TableMappingNotFoundException("Unable to resolve mapping for table reference " + tableRef);
+            }
+        }
+        return candidate;
+    }
+
+    protected void validateShortName(TableReference tableRef, TableReference shortName) {
+        Validate.isTrue(Schemas.isTableNameValid(shortName.getQualifiedName()),
+                "Table mapper has an invalid table name for table reference %s: %s", tableRef, shortName);
+    }
+
+
+    @Override
+    public <T> Map<TableReference, T> mapToShortTableNames(Map<TableReference, T> toMap)
+            throws TableMappingNotFoundException {
+        Map<TableReference, T> newMap = Maps.newHashMap();
+        for (Map.Entry<TableReference, T> e : toMap.entrySet()) {
+            newMap.put(getMappedTableName(e.getKey()), e.getValue());
```

Comment:
this is sketch, can we get the map, then get the things from it?

Reply:
The fix here would result in more unreadable code, so given that there is no contract about what should happen in presence of race conditions, I'm leaning to leaving this as is